lib/execution.go

```diff
@@ -0,0 +1,562 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2019 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package lib
```

Comment:
What do you think about moving this and related files to a sub-package (`k6/core/execution` or `k6/core/engine/execution`)? If we want to clean up `lib` we probably shouldn't add more technical debt to that. :)

Unless you meant these to be used across the entire application, though I don't see a need for that.

Reply:
I'm OK with leaving this for later, though having a clean package separation would help with testing, and postponing this might make things difficult to move once the functionality is final (or we might be past that point already).

As for the locations, how about:

- top-level `k6/engine` package (`execution` is a bit vague, and `core` is a synonym for `lib` ;)).
  I know `engine` is "taken" by `core.Engine`, but we should probably move this to `k6/engine` and abandon `k6/core`.
  All interfaces like `Executor`, `ExecutorConfig`, etc. could exist in `k6/engine/interfaces.go`.
- implementations of `interfaces.go` could exist in `k6/engine/execution.go`, `k6/engine/engine.go`, etc.
- `Execution*` helper objects could also be in `k6/engine/execution.go`.

Essentially, I'm not sure we need a separate "execution" package if we think of these new executors as being part of the "k6 engine". But you guys are more familiar with the codebase, so I'll fallback to your decision.

Reply:
> or we might be past that point already

Yeah, we pretty much are - the effort of changing the module structure now or later should roughly be the same, but later we'd have a better idea what and how we want to change. And before merging this in `master`, I'll probably spend a day to squash everything and split it into sensible commits anyway.

Regarding the modules:
- :+1: fully agree with abandoning `core` and it beeing not much better than `lib`, it doesn't communicate very clearly what it does. Besides, it's very debatable what the "core" of a project like k6 actually is, but making and measuring of HTTP requests would probably be a better fit than the current engine.
- The `engine` actually doesn't have much to do with execution, either before or after this PR. Open the code and you'll see that it mostly has to do with metric collecting and processing (outputs, thresholds, summary, etc.). So it should likely stay separate. The `local` `ExecutionScheduler` on the other hand is very much connected and I don't see and reason not to be in the same place as the things from point 3. in my list above. Its current imports are:
   ```
   github.com/loadimpact/k6/lib
   github.com/loadimpact/k6/stats
   github.com/loadimpact/k6/ui/pb
   ```
- So, I think we should keep the `engine` and `execution`-related things separate, in their own top-level modules `k6/engine` and `k6/execution`. And I disagree that `execution` is vague - it would handle how the user-specified `execution` option is processed and it would contain code that determines how k6 executes the users' scripts according to it (or the execution shortcut options). The only potential alternative was `scheduler`/`scheduling`, and that was what it used to be called, but we renamed it to `execution` because of the reasons listed in https://github.com/loadimpact/k6/pull/1007#issuecomment-497319205